### DISCUSSION TITLE
is it a right BANK2_OFFSET define in flash_stm32l5x.c?

### DISCUSSION THREAD

**Discussion by liweifeng81** [zephyrproject-rtos/zephyr#65712] [Q&A]
the below bold code is for avoiding the range is crossing bank. but the define of BANK2_OFFSET is not right.
e.g. I'm using stm32u545, the flash size is 512k, and BANK2_OFFSET should be 256k. 
I think it should not use BANK2_OFFSET, but FLASH_SIZE / 2 instead.

#if defined(CONFIG_SOC_SERIES_STM32H5X)
/* at this time stm32h5 mcus have 128KB (stm32h50x) or 2MB (stm32h56x/57x) */
#define STM32_SERIES_MAX_FLASH	2048
#elif defined(CONFIG_SOC_SERIES_STM32L5X)
#define STM32_SERIES_MAX_FLASH	512
#elif defined(CONFIG_SOC_SERIES_STM32U5X)
/* at this time stm32u5 mcus have 1MB (stm32u575) or 2MB (stm32u585) */
#define STM32_SERIES_MAX_FLASH	2048
#endif

#define PAGES_PER_BANK ((FLASH_SIZE / FLASH_PAGE_SIZE) / 2)

#define BANK2_OFFSET	(KB(STM32_SERIES_MAX_FLASH) / 2)

bool flash_stm32_valid_range(const struct device *dev, off_t offset,
			     uint32_t len,
			     bool write)
{
	if (stm32_flash_has_2_banks(dev) &&
			(CONFIG_FLASH_SIZE < STM32_SERIES_MAX_FLASH)) {
		/*
		 * In case of bank1/2 discontinuity, the range should not
		 * start before bank2 and end beyond bank1 at the same time.
		 * Locations beyond bank2 are caught by
		 * flash_stm32_range_exists.
		 */
		**if ((offset < BANK2_OFFSET) &&
					(offset + len > FLASH_SIZE / 2))** {
			return 0;
		}
	}

	return (!write || (offset % 8 == 0 && len % 8 == 0U)) &&
		flash_stm32_range_exists(dev, offset, len);
}

**Comment by erwango** [comment]
@liweifeng81 Looks you're right. When this code has been implemented we didn't support as much as flash variant as today.
Can you create a PR?

**Reply by liweifeng81** [reply]
Thanks @erwango for your quick reply. I will study how to create a PR.
and  there's another issue on the same file blocking me. in function "flash_stm32_page_layout", there's the following code.
the "Dummy page" seems not on stm32u57x according to ST's spec RM0456, and also for stm32l552xx according to RM0438.
If I remove the "Dummy page" and change the array size of "stm32_flash_layout" to 2, it works in my project.
Can you help me check it as well?

	if (stm32_flash_has_2_banks(dev) &&
			(CONFIG_FLASH_SIZE < STM32_SERIES_MAX_FLASH)) {
		/*
		 * For stm32l552xx with 256 kB flash or stm32u57x with 1MB flash
		 * which have space between banks 1 and 2.
		 */

		/* Bank1 */
		stm32_flash_layout[0].pages_count = PAGES_PER_BANK;
		stm32_flash_layout[0].pages_size = FLASH_PAGE_SIZE;

		/* Dummy page corresponding to space between banks 1 and 2 */
		stm32_flash_layout[1].pages_count = 1;
		stm32_flash_layout[1].pages_size = BANK2_OFFSET
				- (PAGES_PER_BANK * FLASH_PAGE_SIZE);

		/* Bank2 */
		stm32_flash_layout[2].pages_count = PAGES_PER_BANK;
		stm32_flash_layout[2].pages_size = FLASH_PAGE_SIZE;

		stm32_flash_layout_size = ARRAY_SIZE(stm32_flash_layout);

**Reply by erwango** [reply]
It might be that your case should fall in the `else` part with `CONFIG_FLASH_SIZE = STM32_SERIES_MAX_FLASH`.

Can you check if fixing `STM32_SERIES_MAX_FLASH` wouldn't fix both issues at same time ?

**Reply by liweifeng81** [reply]
you are right, it works good by changing STM32_SERIES_MAX_FLASH to 512(CONFIG_FLASH_SIZE).

I changed BANK2_OFFSET to FLASH_SIZE / 2 as the **bold** code in my original post.

I'm not sure whether I misunderstood the design.
If I change STM32_SERIES_MAX_FLASH to 512, it's not check "bank1/2 discontinuity" case and this "stm32_flash_layout" become 1. actually the bank 1 and 2 are continuous on both STM32U535/545 and STM32U575/585 according to Table51/52 in RM0456. so it should be OK. But that will conflict with the feature "Bank swapping" and Trust Zone, but the 2 feature is not supported yet in Zephyr, am I right?

So a simple fix is just define STM32_SERIES_MAX_FLASH as CONFIG_FLASH_SIZE, then the "BANK2_OFFSET" is also right. 
... 
#elif defined(CONFIG_SOC_SERIES_STM32U5X)
/* at this time stm32u5 mcus have 1MB (stm32u575) or 2MB (stm32u585) */
#define STM32_SERIES_MAX_FLASH	**(CONFIG_FLASH_SIZE)**
#endif
...
